pixeltable/functions/together.py

```diff
@@ -1,27 +1,121 @@
         max_tokens: Optional[int] = None,
-        repetition_penalty: Optional[float] = None,
         stop: Optional[list] = None,
-        top_k: Optional[int] = None,
+        temperature: Optional[float] = None,
         top_p: Optional[float] = None,
-        temperature: Optional[float] = None
+        top_k: Optional[int] = None,
+        repetition_penalty: Optional[float] = None,
+        logprobs: Optional[int] = None,
+        echo: Optional[bool] = None,
+        n: Optional[int] = None,
+        safety_model: Optional[str] = None
 ) -> dict:
-    import together
-    return together.Complete.create(
-        prompt,
-        model,
+    return together_client().completions.create(
+        prompt=prompt,
+        model=model,
         max_tokens=max_tokens,
-        repetition_penalty=repetition_penalty,
         stop=stop,
+        temperature=temperature,
+        top_p=top_p,
         top_k=top_k,
+        repetition_penalty=repetition_penalty,
+        logprobs=logprobs,
+        echo=echo,
+        n=n,
+        safety_model=safety_model
+    ).dict()
+
+
+@pxt.udf
+def chat_completions(
+        messages: list[dict[str, str]],
+        *,
+        model: str,
+        max_tokens: Optional[int] = None,
+        stop: Optional[list[str]] = None,
+        temperature: Optional[float] = None,
+        top_p: Optional[float] = None,
+        top_k: Optional[int] = None,
+        repetition_penalty: Optional[float] = None,
+        logprobs: Optional[int] = None,
+        echo: Optional[bool] = None,
+        n: Optional[int] = None,
+        safety_model: Optional[str] = None,
+        response_format: Optional[dict] = None,
+        tools: Optional[dict] = None,
+        tool_choice: Optional[dict] = None
+) -> dict:
+    return together_client().chat.completions.create(
+        messages=messages,
+        model=model,
+        max_tokens=max_tokens,
+        stop=stop,
+        temperature=temperature,
         top_p=top_p,
-        temperature=temperature
+        top_k=top_k,
+        repetition_penalty=repetition_penalty,
+        logprobs=logprobs,
+        echo=echo,
+        n=n,
+        safety_model=safety_model,
+        response_format=response_format,
+        tools=tools,
+        tool_choice=tool_choice
+    ).dict()
+
+
+@pxt.udf(batch_size=32, return_type=pxt.ArrayType((None,), dtype=pxt.FloatType()))
+def embeddings(input: Batch[str], *, model: str) -> Batch[np.ndarray]:
+    result = together_client().embeddings.create(input=input, model=model)
+    return [
+        np.array(data.embedding, dtype=np.float64)
+        for data in result.data
+    ]
+
+
+@pxt.udf
+def image_generations(
+        prompt: str,
+        *,
+        model: str,
+        steps: Optional[int] = None,
+        seed: Optional[int] = None,
+        height: Optional[int] = None,
+        width: Optional[int] = None,
+        negative_prompt: Optional[str] = None,
+) -> PIL.Image.Image:
+    result = together_client().images.generate(
```

Comment:
What is the type of result? It would be good to decompose this into the endpoint call and the post-processing, in the form of an expr udf.